volatility3/framework/symbols/linux/extensions/__init__.py

```diff
@@ -2410,3 +2410,107 @@ def get_nodes(self) -> Iterator[int]:
+        # Depends on CONFIG_NEED_SG_DMA_LENGTH
+        if self.has_member("dma_length"):
+            return self.dma_length
+        return self.length
+
+    def _get_sg_max_single_alloc(self) -> int:
+        """Based on kernel's SG_MAX_SINGLE_ALLOC.
+
+        Doc. from kernel source :
+            * Maximum number of entries that will be allocated in one piece, if
+            * a list larger than this is required then chaining will be utilized.
+        """
+        return self._context.layers[self.vol.layer_name].page_size // self.vol.size
+
+    def _sg_next(self) -> interfaces.objects.ObjectInterface:
+        """Get the next scatterlist struct from the list.
+        Based on kernel's sg_next.
+
+        Doc. from kernel source :
+            * Notes on SG table design.
+            *
+            * We use the unsigned long page_link field in the scatterlist struct to place
+            * the page pointer AND encode information about the sg table as well. The two
+            * lower bits are reserved for this information.
+            *
+            * If bit 0 is set, then the page_link contains a pointer to the next sg
+            * table list. Otherwise the next entry is at sg + 1.
+            *
+            * If bit 1 is set, then this sg entry is the last element in a list.
+        """
+        if self._sg_is_last():
+            return None
+
+        if self._sg_is_chain():
+            next_address = self._sg_chain_ptr()
+        else:
+            next_address = self.vol.offset + self.vol.size
+
+        sg = self._context.object(
+            self.get_symbol_table_name() + constants.BANG + "scatterlist",
+            self.vol.layer_name,
+            next_address,
+        )
+        return sg
+
+    def for_each_sg(self) -> Iterator[interfaces.objects.ObjectInterface]:
+        """Iterate over each struct in the scatterlist."""
+        sg = self
+        sg_max_single_alloc = self._get_sg_max_single_alloc()
+
+        # Empty scatterlists protection
+        if sg.page_link == 0 and sg._sg_dma_len() == 0 and sg.dma_address == 0:
+            return None
+        else:
+            # Yield itself first
+            yield sg
+
+        entries_count = 1
+        # entries_count <= sg_max_single_alloc should always be true if the
+        # scatterlists were correctly chained.
+        while entries_count <= sg_max_single_alloc:
+            sg = sg._sg_next()
+            if sg is None:
+                break
+            # Points to a new scatterlist
+            elif sg._sg_is_chain():
+                entries_count = 0
+            else:
+                entries_count += 1
+                yield sg
+
+    def get_content(
+        self,
+    ) -> Iterator[bytes]:
+        """Traverse a scatterlist to gather content located at each
+        dma_address position.
+
+        Returns:
+            An iterator of bytes
+        """
+        physical_layer = self._context.layers["memory_layer"]
```

Comment:
So this is... fragile.  For Intel layers, the underlying layer will likely be `memory_layer`  (but not in the case of a virtualized intel system, for instance), and there's no telling that Arm's lower layer will be called `memory_layer`.  I don't immediately have a better way of doing this.  The layer's dependencies list would tell you all the layers it lives on top of, and they're *likely* to be in order, but short of mapping an address down a layer (and if you pick the wrong one, you might hit swap or something else) there's no guaranteed way of knowing how the layers all fit together...  Any thoughts how best to handle this?

Reply:
What we want to achieve, is basically getting `kernel` OR `process_layer` - 1 (base) layer, which should in most cases be the RAM. 

Maybe something like `self._context.layers[self.vol.layer_name].base_layer` should do it ?

Reply:
Ok, so here it can be a bit tricky, as we are "blind" of the layer position in the stacking list. However, pdbscan, linux.py and mac.py will refer to the memory layer by the config key "memory_layer". We can use this information to determine where we are.

Consequently, this implementation should do the trick :

```python 
# Either memory is layer-1 because we are a module layer, either memory is the current layer
memory_layer_name = self.context.layers[self.vol.layer_name].config.get(
    "memory_layer", self.vol.layer_name
)
memory_layer = self.context.layers[memory_layer_name]
```

Reply:
Yep, definitely need to use the code above to get it from this object's layer.  At the moment all plugins only deal with one image, but that's absolutely not the intention long term.

Reply:
Change has been applied.